File: ramenmobileapp/lib/services/paymongo_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:url_launcher/url_launcher.dart';

class PayMongoService {
  static const String _baseUrl = 'http://192.168.0.106:3000/api/v1/paymongo'; // Correct path with /v1
  
  // Get authentication token
  static Future<String?> _getToken() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      return prefs.getString('auth_token');
    } catch (e) {
      print('Error getting token: $e');
      return null;
    }
  }
  
  // Method 1: Direct Redirect Payment (No QR Code - Automation Friendly)
  static Future<Map<String, dynamic>> createDirectPayment({
    required double amount,
    required String orderId,
    required String paymentMethod, // 'gcash' or 'paymaya'
  }) async {
    try {
      print('🔄 Creating direct $paymentMethod payment for ₱$amount (Order: $orderId)');
      
      final token = await _getToken();
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$_baseUrl/create-direct-payment'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'amount': amount,
          'orderId': orderId,
          'paymentMethod': paymentMethod,
          'successUrl': 'ramenxpress://payment/success',
          'failureUrl': 'ramenxpress://payment/failed',
        }),
      );

      print('🔍 Response status: ${response.statusCode}');
      print('🔍 Response body: ${response.body}');
      
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          print('✅ Direct payment created successfully');
          return {
            'success': true,
            'sourceId': data['data']['sourceId'],
            'checkoutUrl': data['data']['checkoutUrl'], 
            'amount': data['data']['amount'],
            'orderId': data['data']['orderId'],
            'paymentMethod': data['data']['paymentMethod'],
          };
        } else {
          print('❌ Backend returned success=false: ${data['message']}');
          throw Exception(data['message'] ?? 'Failed to create payment');
        }
      } else {
        print('❌ HTTP Error ${response.statusCode}: ${response.body}');
        try {
          final errorData = jsonDecode(response.body);
          throw Exception(errorData['message'] ?? 'Server error: ${response.statusCode}');
        } catch (e) {
          throw Exception('Server error: ${response.statusCode} - ${response.body}');
        }
      }
    } catch (e) {
      print('❌ Direct payment creation error: $e');
      return {
        'success': false,
        'error': e.toString(),
      };
    }
  }

  // Method 2: Payment Intent with Test Cards (Best for Automation)
  static Future<Map<String, dynamic>> createPaymentIntent({
    required double amount,
    required String orderId,
    String description = 'RamenXpress Order',
  }) async {
    try {
      print('🔄 Creating payment intent for ₱$amount (Order: $orderId)');
      
      final token = await _getToken();
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$_baseUrl/create-payment-intent'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'amount': amount,
          'orderId': orderId,
          'description': description,
          'paymentMethods': ['card', 'gcash', 'paymaya'], // Multiple options
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          print('✅ Payment intent created successfully');
          return {
            'success': true,
            'paymentIntentId': data['data']['paymentIntentId'],
            'clientKey': data['data']['clientKey'],
            'checkoutUrl': data['data']['checkoutUrl'],
            'amount': data['data']['amount'],
            'orderId': data['data']['orderId'],
          };
        } else {
          throw Exception(data['message'] ?? 'Failed to create payment intent');
        }
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['message'] ?? 'Server error: ${response.statusCode}');
      }
    } catch (e) {
      print('❌ Payment intent creation error: $e');
      return {
        'success': false,
        'error': e.toString(),
      };
    }
  }

  // Method 3: Simulate Payment (Development/Testing Only)
  static Future<Map<String, dynamic>> simulatePayment({
    required double amount,
    required String orderId,
    required String paymentMethod,
    bool shouldSucceed = true, // For testing different outcomes
  }) async {
    try {
      print('🧪 Simulating $paymentMethod payment for ₱$amount (Order: $orderId)');
      
      final token = await _getToken();
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.post(
        Uri.parse('$_baseUrl/simulate-payment'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode({
          'amount': amount,
          'orderId': orderId,
          'paymentMethod': paymentMethod,
          'shouldSucceed': shouldSucceed,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          print('✅ Payment simulation completed');
          return {
            'success': true,
            'paymentId': data['data']['paymentId'],
            'status': data['data']['status'],
            'amount': data['data']['amount'],
            'orderId': data['data']['orderId'],
            'paymentMethod': data['data']['paymentMethod'],
            'isSimulated': true,
          };
        } else {
          throw Exception(data['message'] ?? 'Payment simulation failed');
        }
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['message'] ?? 'Server error: ${response.statusCode}');
      }
    } catch (e) {
      print('❌ Payment simulation error: $e');
      return {
        'success': false,
        'error': e.toString(),
      };
    }
  }

  // Open payment URL in browser/webview (User-friendly with fallback)
  static Future<bool> openPaymentUrl(String checkoutUrl) async {
    try {
      print('🔗 Attempting to open payment URL: $checkoutUrl');
      
      final uri = Uri.parse(checkoutUrl);
      
      // Try to check if URL can be launched
      bool canLaunch = false;
      try {
        canLaunch = await canLaunchUrl(uri);
      } catch (e) {
        print('⚠️ Cannot check URL launch capability: $e');
        canLaunch = true; // Assume it can be launched
      }
      
      if (canLaunch) {
        try {
          await launchUrl(
            uri,
            mode: LaunchMode.externalApplication, // Opens in external browser
          );
          print('✅ Payment URL opened successfully');
          return true;
        } catch (e) {
          print('⚠️ Failed to launch URL with external app, trying platform default: $e');
          try {
            await launchUrl(uri, mode: LaunchMode.platformDefault);
            print('✅ Payment URL opened with platform default');
            return true;
          } catch (e2) {
            print('❌ All URL launch methods failed: $e2');
            return false;
          }
        }
      } else {
        print('❌ Cannot launch payment URL: $checkoutUrl');
        return false;
      }
    } catch (e) {
      print('❌ Error opening payment URL: $e');
      return false;
    }
  }

  // Verify payment status (works with all methods)
  static Future<Map<String, dynamic>> verifyPayment(String paymentId) async {
    try {
      print('🔍 Verifying payment status for: $paymentId');
      
      final token = await _getToken();
      if (token == null) {
        throw Exception('Authentication token not found');
      }

      final response = await http.get(
        Uri.parse('$_baseUrl/verify-payment/$paymentId'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success']) {
          final status = data['data']['status'];
          print('📊 Payment status: $status');
          
          return {
            'success': true,
            'id': data['data']['id'],
            'status': status,
            'amount': data['data']['amount'],
            'currency': data['data']['currency'],
            'isPaid': status == 'paid' || status == 'succeeded' || status == 'chargeable',
            'isPending': status == 'pending' || status == 'awaiting_payment_method',
            'isFailed': status == 'failed' || status == 'cancelled',
          };
        } else {
          throw Exception(data['message'] ?? 'Payment verification failed');
        }
      } else {
        final errorData = jsonDecode(response.body);
        throw Exception(errorData['message'] ?? 'Server error: ${response.statusCode}');
      }
    } catch (e) {
      print('❌ Payment verification error: $e');
      return {
        'success': false,
        'error': e.toString(),
      };
    }
  }

  // Get test card information for development
  static Map<String, dynamic> getTestCards() {
    return {
      'success_card': {
        'number': '4242424242424242',
        'expiry': '12/25',
        'cvc': '123',
        'description': 'Always succeeds',
      },
      'decline_card': {
        'number': '4000000000000002',
        'expiry': '12/25',
        'cvc': '123',
        'description': 'Always declines',
      },
      'auth_required_card': {
        'number': '4000000000009995',
        'expiry': '12/25',
        'cvc': '123',
        'description': 'Requires authentication',
      },
    };
  }

  // Check if payment method is supported
  static bool isPaymentMethodSupported(String paymentMethod) {
    return ['gcash', 'paymaya', 'maya', 'card'].contains(paymentMethod.toLowerCase());
  }

  // Get minimum payment amount
  static double getMinimumAmount() {
    return 20.0; // PayMongo requires minimum ₱20
  }

  // Format amount for display
  static String formatAmount(double amount) {
    return '₱${amount.toStringAsFixed(2)}';
  }

  // Generate order ID for PayMongo
  static String generateOrderId() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    return 'MOBILE-$timestamp';
  }
}

File: backend/services/paymongoService.js
const axios = require('axios');
const QRCode = require('qrcode');

class PayMongoService {
    constructor() {
        this.baseURL = 'https://api.paymongo.com/v1';
        this.secretKey = process.env.PAYMONGO_SECRET_KEY;
        
        if (!this.secretKey) {
            console.warn('⚠️  PayMongo secret key not found. Please set PAYMONGO_SECRET_KEY in your environment variables.');
        }
    }

    // Generate QR code for POS payment using Sources API
    async generatePOSQRCode(amount, orderId, paymentMethod = 'gcash', frontendUrl = null) {
        try {
            console.log(`🔄 Generating ${paymentMethod} QR code for ₱${amount} (Order: ${orderId})`);

            // Map payment method to PayMongo source type
            const sourceTypeMap = {
                'gcash': 'gcash',
                'paymaya': 'gcash', // PayMaya uses same source type as GCash
                'grabpay': 'grab_pay'
            };

            const sourceType = sourceTypeMap[paymentMethod] || 'gcash';

            // Create source for QR code
            const response = await axios.post(`${this.baseURL}/sources`, {
                data: {
                    attributes: {
                        amount: Math.round(amount * 100), // Convert to centavos
                        currency: 'PHP',
                        type: sourceType,
                        redirect: {
                            success: `${frontendUrl || process.env.FRONTEND_URL || 'http://127.0.0.1:5501'}/frontend/html/pos.html?payment=success`,
                            failed: `${frontendUrl || process.env.FRONTEND_URL || 'http://127.0.0.1:5501'}/frontend/html/pos.html?payment=failed`
                        }
                    }
                }
            }, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });

            const source = response.data.data;
            console.log(`✅ Source created: ${source.id}`);

            // Get redirect URL for QR code
            const redirectUrl = source.attributes.redirect.checkout_url;
            console.log(`✅ Redirect URL: ${redirectUrl}`);

            // Generate QR code
            const qrCodeDataURL = await QRCode.toDataURL(redirectUrl, {
                width: 400, // Larger for POS display
                margin: 2,
                color: {
                    dark: '#000000',
                    light: '#FFFFFF'
                }
            });

            console.log(`✅ QR code generated for ${paymentMethod}`);

            return {
                success: true,
                data: {
                    sourceId: source.id,
                    redirectUrl: redirectUrl,
                    qrCodeDataURL: qrCodeDataURL,
                    amount: amount,
                    orderId: orderId,
                    paymentMethod: paymentMethod,
                    status: source.attributes.status,
                    createdAt: new Date()
                }
            };

        } catch (error) {
            console.error('❌ PayMongo QR generation error:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data?.errors?.[0]?.detail || error.message
            };
        }
    }

    // Generate direct payment for mobile app (no QR code)
    async generateMobileDirectPayment(amount, orderId, paymentMethod = 'gcash') {
        try {
            console.log(`🔄 Generating mobile ${paymentMethod} payment for ₱${amount} (Order: ${orderId})`);
            
            // Check if secret key is available
            if (!this.secretKey) {
                console.error('❌ PayMongo secret key not configured');
                return {
                    success: false,
                    error: 'PayMongo secret key not configured. Please set PAYMONGO_SECRET_KEY in environment variables.'
                };
            }
            
            console.log('🔑 PayMongo secret key is configured');

            // Map payment method to PayMongo source type
            const sourceTypeMap = {
                'gcash': 'gcash',
                'paymaya': 'gcash', // PayMaya uses same source type as GCash
                'grabpay': 'grab_pay'
            };

            const sourceType = sourceTypeMap[paymentMethod] || 'gcash';

            const requestData = {
                data: {
                    attributes: {
                        amount: Math.round(amount * 100), // Convert to centavos
                        currency: 'PHP',
                        type: sourceType,
                        redirect: {
                            success: 'http://192.168.0.106:3000/api/v1/paymongo/mobile-payment-success',
                            failed: 'http://192.168.0.106:3000/api/v1/paymongo/mobile-payment-failed'
                        }
                    }
                }
            };
            
            console.log('📤 PayMongo API request:', JSON.stringify(requestData, null, 2));
            
            // Create source for mobile payment with simple success page
            const response = await axios.post(`${this.baseURL}/sources`, requestData, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });
            
            console.log('📥 PayMongo API response status:', response.status);
            console.log('📥 PayMongo API response data:', JSON.stringify(response.data, null, 2));

            const source = response.data.data;
            console.log(`✅ Mobile source created: ${source.id}`);

            // Get redirect URL for mobile payment
            const redirectUrl = source.attributes.redirect.checkout_url;
            console.log(`✅ Mobile redirect URL: ${redirectUrl}`);

            return {
                success: true,
                data: {
                    sourceId: source.id,
                    redirectUrl: redirectUrl,
                    amount: amount,
                    orderId: orderId,
                    paymentMethod: paymentMethod,
                    status: source.attributes.status,
                    createdAt: new Date()
                }
            };

        } catch (error) {
            console.error('❌ PayMongo mobile payment error:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data?.errors?.[0]?.detail || error.message
            };
        }
    }

    // Verify payment intent status
    async verifyPaymentIntent(paymentIntentId) {
        try {
            const response = await axios.get(`${this.baseURL}/payment_intents/${paymentIntentId}`, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });

            return {
                success: true,
                data: response.data.data
            };
        } catch (error) {
            console.error('PayMongo verifyPaymentIntent error:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data || error.message
            };
        }
    }

    // Legacy method for backward compatibility
    async verifyPaymentSource(sourceId) {
        // If it looks like a payment intent ID, use the new method
        if (sourceId && sourceId.startsWith('pi_')) {
            return await this.verifyPaymentIntent(sourceId);
        }
        
        try {
            const response = await axios.get(`${this.baseURL}/sources/${sourceId}`, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });

            return {
                success: true,
                data: response.data.data
            };
        } catch (error) {
            console.error('PayMongo verifyPaymentSource error:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data || error.message
            };
        }
    }

    // Charge a chargeable payment source
    async chargePaymentSource(sourceId, amount) {
        try {
            console.log(`💳 Charging payment source: ${sourceId} for ₱${amount}`);

            const response = await axios.post(`${this.baseURL}/payments`, {
                data: {
                    attributes: {
                        amount: Math.round(amount * 100), // Convert to centavos
                        currency: 'PHP',
                        source: {
                            id: sourceId,
                            type: 'source'
                        }
                    }
                }
            }, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });

            console.log(`✅ Payment charged successfully: ${response.data.data.id}`);
            return {
                success: true,
                data: response.data.data
            };
        } catch (error) {
            console.error('❌ Payment charging failed:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data || error.message
            };
        }
    }

    // Create payment intent for card payments (if needed later)
    async createPaymentIntent(amount, currency = 'PHP', description = 'RamenXpress POS Order') {
        try {
            const response = await axios.post(`${this.baseURL}/payment_intents`, {
                data: {
                    attributes: {
                        amount: Math.round(amount * 100),
                        currency: currency,
                        description: description,
                        payment_method_allowed: ['card', 'paymaya', 'gcash']
                    }
                }
            }, {
                headers: {
                    'Authorization': `Basic ${Buffer.from(this.secretKey + ':').toString('base64')}`,
                    'Content-Type': 'application/json'
                }
            });

            return {
                success: true,
                data: response.data.data
            };
        } catch (error) {
            console.error('PayMongo createPaymentIntent error:', error.response?.data || error.message);
            return {
                success: false,
                error: error.response?.data || error.message
            };
        }
    }
}

module.exports = new PayMongoService();
